### ISSUE TITLE
Nutrition Label Ex 35-36: Should use Flexbox, not Float

### ISSUE THREAD

**Issue by SaintPeter** [freeCodeCamp/freeCodeCamp#46675]
### Describe the Issue

I was helping a camper with this challenge today and they were confused about the need to add `overflow: hidden` to the `h1` element.  I was too, until I did some research and remembered that [containers don't clear floats automatically](https://css-tricks.com/containers-dont-clear-floats/) unless you use `overflow: hidden`.  This is very counter-intuitive and is not explained at all in the challenge.

Additionally, in the next challenge you have to add a "magic number" to the h1 to get it to optically align. I don't think this scales well (literally), because if you increase your font size it's going to break.

### Affected Page

https://www.freecodecamp.org/learn/2022/responsive-web-design/learn-typography-by-building-a-nutrition-label/step-35

### Your code

```html
<div class="calories-info">
        <div class="left-container">
          <p class="bold sm-text">Amount per serving</p>
            <h1>Calories</h1>
          </div>
        <h1 class="right">230</h1>
      </div>          
```
```css
.calories-info {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
}

.left-container {
  display: flex;
  flex-direction: column;
}
```

### Expected behavior

It seems to me that there is a better, much cleaner solution to be had by applying another wrapper around the "Amount Per Serving" and "Calories" elements and using a flex column, then using flex on the `.calories-info` div with `justify-content: space-between` and `align-items: flex-end`;  

No magic numbers needed, both the markup and css are cleaner - although you do need one more containing element.

I've added a screenshot which shows the layout of the elements in my example.

### Screenshots

![image](https://user-images.githubusercontent.com/553494/175826124-d65a4ab3-7209-479c-9ed9-f7918db36916.png)

### System

N/A

### Additional context

In general, the given solution feels very "old school" to me.  Flexbox has been available on most platforms for a number of years now.  We really shouldn't be teaching people to support IE11 (especially since it has reached the end of it's support life, recently).

I know this is a big change, it just bugs me to see this being taught in what is supposed to be a modern course on the web.

**Comment by Sboonny** [comment]
At the final step, it looks the same except the font in `.calories-info span` aren't reading 

```css
.calories-info span {
  font-size: 1.2em;
}
```

anyway, I like the idea of redoing the challenge and polishing it, maybe even adding hints about grid concepts.


![image](https://user-images.githubusercontent.com/88248797/175827402-00646986-2d99-4978-a120-35b337b7a4ea.png)


here is the code, because I may have made a mistake.

<details>
 <summary> html</summary>
 
```html
 <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>Nutrition Label</title>
    <link href="https://fonts.googleapis.com/css?family=Open+Sans:400,700,800" rel="stylesheet">
    <link href="./styles.css" rel="stylesheet">
  </head>
  <body>
    <div class="label">
      <header>
        <h1 class="bold">Nutrition Facts</h1>
        <div class="divider"></div>
        <p>8 servings per container</p>
        <p class="bold">Serving size <span class="right">2/3 cup (55g)</span></p>
      </header>
      <div class="divider lg"></div>
      <div class="calories-info">
        <div class="left-container">
        <p class="bold sm-text">Amount per serving</p>
        <h1>Calories </h1>
      </div>
         <span class="right">230</span>

      </div>
      <div class="divider md"></div>
      <div class="daily-value sm-text">
        <p class="right bold no-divider">% Daily Value *</p>
        <div class="divider"></div>
        <p><span class="bold">Total Fat</span> 8g <span class="bold right">10%</span></p>
        <p class="indent no-divider">Saturated Fat 1g <span class="bold right">5%</span></p>
        <div class="divider"></div>
        <p class="indent no-divider"><i>Trans</i> Fat 0g</p>
        <div class="divider"></div>
        <p><span class="bold">Cholesterol</span> 0mg <span class="right bold">0%</span></p>
        <p><span class="bold">Sodium</span> 160mg <span class="right bold">7%</span></p>
        <p><span class="bold">Total Carbohydrate</span> 37g <span class="right bold">13%</span></p>
        <p class="indent no-divider">Dietary Fiber 4g</p>
        <div class="divider"></div>
        <p class="indent no-divider">Total Sugars 12g</p>
        <div class="divider dbl-indent"></div>
        <p class="dbl-indent no-divider">Includes 10g Added Sugars <span class="right bold">20%</span>
        <div class="divider"></div>
        <p class="no-divider"><span class="bold">Protein</span> 3g</p>
        <div class="divider lg"></div>
        <p>Vitamin D 2mcg <span class="right">10%</span></p>
        <p>Calcium 260mg <span class="right">20%</span></p>
        <p>Iron 8mg <span class="right">45%</span></p>
        <p class="no-divider">Potassium 235mg <span class="right">6%</span></p>
      </div>
      <div class="divider md"></div>
      <p class="note">* The % Daily Value (DV) tells you how much a nutrient in a serving of food contributes to a daily diet. 2,000 calories a day is used for general nutrition advice.</p>
    </div>
  </body>
</html>
```
 </details>

<details>
 <summary> css</summary>
 
```css
* {
  box-sizing: border-box;
}

html {
  font-size: 16px;
}

body {
  font-family: 'Open Sans', sans-serif;
}

.label {
  border: 2px solid black;
  width: 270px;
  margin: 20px auto;
  padding: 0 7px;
}

header h1 {
  text-align: center;
  margin: -4px 0;
  letter-spacing: 0.15px
}

p {
  margin: 0;
}

.divider {
  border-bottom: 1px solid #888989;
  margin: 2px 0;
  clear: right;
}

.bold {
  font-weight: 800;
}

.right {
  float: right;
}

.lg {
  height: 10px;
}

.lg, .md {
  background-color: black;
  border: 0;
}

.md {
  height: 5px;
}

.sm-text {
  font-size: 0.85rem;
}

.calories-info {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
}

.left-container {
  display: flex;
  flex-direction: column;
}

.calories-info h1 {
  margin: -5px -2px;
}

.calories-info span {
  font-size: 1.2em;
}

.indent {
  margin-left: 1em;
}

.dbl-indent {
  margin-left: 2em;
}

.daily-value p:not(.no-divider) {
  border-bottom: 1px solid #888989;
}

.note {
  font-size: 0.6rem;
  margin: 5px 0;
}
```
 </details>

**Comment by nhcarrigan** [comment]
@Sboonny We would not want to introduce `grid` here, as that's not taught until later in the curriculum.

However, this project does come immediately after the flexbox project, and would be a good opportunity to reinforce those concepts.

**Comment by RandellDawson** [comment]
Where will float be taught?  It was decided over 2 years ago that we still needed to teach float so when users come across it in other code, they will understand how it works.

**Comment by SaintPeter** [comment]
I'm not where where it should be taught.  It should be taught the way it is currently intended to be used: for inline elements that need text to flow around them.  A news article with an image on the side which the text wraps around is ideal.  The way it was historically used (like in this exercise) is kinda a bastardization of it's intended function.

**Comment by bbsmooth** [comment]
I'll throw a wrench into this discussion and make a case for using float in this heading (at least the way it is currently marked up). And of course because it's me you have to know that it's accessibility related. Using the NVDA screen reader, when navigating the page by up/down arrow, if you use flex to align "Calories" and "230" then it will read them as separate `h1` headings. For example, if you arrow key down into the heading then NVDA will only announce "Calories, heading level 1". And if you do a second down arrow then it will announce "230, heading level 1". Apparently NVDA treats these as if they are two separate elements since they are being displayed with flex. But when using the current float technique, NVDA always announces them together as one heading. 

Of course you could just write this off as a buggy implementation in NVDA. But would it really hurt to continue to use float for this one heading in order to appease NVDA (and possibly other screen readers, I haven't tested them all)? 

Actually, I think the best solution here is to not include the number in the heading at all. The heading is just "Calories" and then the amount of calories is content after the heading. Then you could use flex to align these and the NVDA issue is solved as well. I think the fact that we are aligning these on the same line and in the same font is giving the false impression that they both need to be in the heading.

**Comment by RandellDawson** [comment]
`Side note: Also, there are two h1s in this markup. While not technically an accessibility violation it is not a best practice and and I think we actually note this in a previous course.`

Yes we do.

**Comment by ieahleen** [comment]
The steps to change would probably be from 23 onward, particularly:

- [ ] step 23 float is used for the first time for class `.right`
- [ ] step 25 makes h1 selector more specific, not needed if there is only one h1
- [ ] step 32 creates a new h1 - it should be a h2 and see https://github.com/freeCodeCamp/freeCodeCamp/issues/46675#issuecomment-1192850012 for the content
- [ ] step 33 styles the second h1
- [ ] step 35 overflow hidden to fix float issue & styles the second h1
- [ ] step 39 used adding class `.right`
- [ ] step 40 fixes float issue
- [ ] step 41 clear property to fix float issue
- [ ] step 42 class right added to an element
- [ ] step 43 same
- [ ] step 52 same
- [ ] step 53 same

and the seed for all from 24 onward, based on previous changes

it's practically redoing the project

**Comment by SaintPeter** [comment]
I'm less concerned about the specific markup and more concerned that we use the proper tool for the job. Flexbox has the advantage of not requiring "magic numbers" in the CSS to fix optical alignment issues.

To @bbsmooth’s point, using a h1 to hold the value is probably semantically wrong, which results in a confusing accessibility experience.

A definition list (`dl`,`dt`,`dd`) for the entire table might be more semantically correct (ref: https://stackoverflow.com/a/50313363/1420506), although it undermines some of the goals of this lesson.

**Comment by ieahleen** [comment]
I have opened #47675 with just the solution updated. This one has also the last step included, which https://github.com/freeCodeCamp/freeCodeCamp/issues/46675#issuecomment-1166604399 didn't.

Could someone check it a moment? If it passes general scrutiny, I will work on adjusing all previous steps to that.

**Comment by bbsmooth** [comment]
Sorry, I didn't look at this as closely as I should have the first time. With the way this is currently marked up, I don't think "Calories" should be a heading. The "Nutrition Facts" `h1` heading is good, but the next heading in this is really "Amount per serving", which should be an `h2`. This should be a heading because it describes the rest of the information below it. "Calories" is really just the same as the other nutrients listed, so it doesn't deserve special treatment. It's styled like a heading, but I think the FDA did that just because that's the most important thing people will want to know about the food so they made it stand out. 

Or, if you want "Calories" to be a heading then it should be an `h3`. But then you should probably make the other bolded nutrients (Total Fat, Cholesteral, etc...) `h3`s as well for consistency. And then any sub nutrients underneath those (`Saturated Fat, Trans Fat, etc...) would be `h4`. This is probably what I would recommend with the current markup.

Honestly, this sort of looks like a table to me (at least everything below calories). But I'm not asking to redo this as a table. As long as we are consistent. For example, as I mentioned above, either every nutrient is a heading or none of them are.

**Comment by ieahleen** [comment]
I will rework the headings, thank you @bbsmooth

**Comment by naomi-lgbt** [comment]
> It's styled like a heading, but I think the FDA did that just because that's the most important thing people will want to know about the food so they made it stand out.

And also because nutrition labels are printed and don't need semantic HTML. 😅

**Comment by ieahleen** [comment]
I am working on trying to make the label as compact as it was, leaving the same design
the current
![image](https://user-images.githubusercontent.com/26656284/194521484-e10d632b-17eb-407c-9e0a-e10265561e4d.png)

my replica
![image](https://user-images.githubusercontent.com/26656284/194521570-1882fef8-e08f-40a0-9fb5-41a7eb3f579d.png)

I will manage, evetually

**Comment by ieahleen** [comment]
About `float`, should it be removed completely?
or is it fine to keep it to align the percentages to the right?
![image](https://user-images.githubusercontent.com/26656284/197335430-b79a8c1a-5438-4d08-876e-c8d2053c404e.png)

**Comment by bbsmooth** [comment]
I don't really have a dog in this float fight, but seeing as the project before this one teaches flexbox then perhaps that might be a reason to use flexbox here (in order to reinforce what they just learned). I think it would be a rather easy change. I was able to do it by removing the float and then adding `display: flex` and `justify-content: space-between` to the `p` ruleset. And then I had to wrap the five nutrients with a bold typeface and the number of grams in a `span` so they would act as one unit (and also Trans Fat since part of it is italicized).

But seeing as the primary purpose of this project is typography then maybe this isn't something that **needs** to be done? After all, the floats do work. But if you put a gun to my head and forced me to pick one, I would go with flexbox over floats.
